ir/src/passes/inlining.rs

```diff
@@ -0,0 +1,211 @@
+use std::{collections::BTreeMap, ops::ControlFlow};
+
+use air_pass::Pass;
+//use miden_diagnostics::DiagnosticsHandler;
+
+use crate::{MirGraph, NodeIndex, Operation};
+
+//pub struct Inlining<'a> {
+//     #[allow(unused)]
+//     diagnostics: &'a DiagnosticsHandler,
+//}
+
+pub struct Inlining {}
+
+//impl<'p> Pass for Inlining<'p> {}
+impl Pass for Inlining {
+    type Input<'a> = MirGraph;
+    type Output<'a> = MirGraph;
+    type Error = ();
+
+    fn run<'a>(&mut self, mut ir: Self::Input<'a>) -> Result<Self::Output<'a>, Self::Error> {
+        match self.run_visitor(&mut ir) {
+            ControlFlow::Continue(()) => Ok(ir),
+            ControlFlow::Break(err) => Err(err),
+        }
+    }
+}
+
+// impl<'a> Inlining<'a> {
+//     pub fn new(diagnostics: &'a DiagnosticsHandler) -> Self {
+//         Self { diagnostics }
+//         Self {}
+//     }
+// }
+impl Inlining {
+    pub fn new() -> Self {
+        Self {}
+    }
+    //TODO MIR: Implement inlining pass on MIR
+    // 1. Understand the basics of the previous inlining process
+    // 2. Remove what is done during lowering from AST to MIR (unroll, ...)
+    // 3. Check how it translates to the MIR structure
+    fn run_visitor(&mut self, ir: &mut MirGraph) -> ControlFlow<()> {
+        inline_all(ir);
+        ControlFlow::Continue(())
+    }
+}
+
+fn inline_all(ir: &mut MirGraph) {
+    // Find all definitions
+    for node_uindex in 0..ir.num_nodes() {
+        let def_node_index = NodeIndex(node_uindex);
+        let def_node = ir.node(&def_node_index).clone();
+        if let Operation::Definition(_, _, body) = &def_node.op {
+            // Find all calls in the body
+            for (index_in_body, call_index) in body.iter().enumerate() {
+                inline_call(ir, call_index, &def_node_index, index_in_body);
+            }
+        }
+    }
+}
+
+fn inline_call(
```

Comment:
It isn't clear to me whether this transformation is sound:

* `inline_all` is iterating over what amounts to a snapshot of the body of the outer definition (if I understand correctly).
* But here in `inline_call`, the actual body of the definition is being mutated/replaced (to insert new items at the index at which the call occurred in the outer def). But after splicing ops into the body of the outer def, all subsequent calls to `inline_call` are using stale indices.

Am I misunderstanding what's actually happening in `inline_all`/`inline_call`?

Reply:
We agree with your analysis and are aware of the current implementation's drawbacks. The next version will use a post-order visit coupled with a work stack where we add newly generated Call nodes while inlining.